Added copy button for individual logs

sakib00 opened the pull request:
## Description
Added copy button to every statements on the logs page 

Copies the entire log object:
```
{
  "eventType": "syslog",
  "timestamp": 1747302095459,
  "id": 1002,
  "level": "INFO",
  "category": "API",
  "subCategory": "UserAssetProvisioningMQTTHandler",
  "message": "Adding publish consumer for topic 'provisioning/+/request': handler=UserAssetProvisioningMQTTHandler"
}
```

Enhancement: #1483 


Before:
<img width="1337" alt="image" src="https://github.com/user-attachments/assets/dcf10615-2315-47fd-ba52-77566ed05b76" />


After:
<img width="1129" alt="image" src="https://github.com/user-attachments/assets/2fd8cd4f-be69-4991-b11a-f4737e46f987" />


## Checklist
<!--
  With all these boxes checked this PR conforms to our Definition of Done.
-->

- [ ] 1. Acceptance criteria of the linked issue(s) are met
- [ ] 2. Tests are written and all tests pass
- [ ] 3. Changes are manually tested by you and the reviewer

sakib00 commented on ui/component/or-log-viewer/src/index.ts:
```diff
@@ -442,6 +452,18 @@ export class OrLogViewer extends translate(i18next)(LitElement) {
+    protected async _copyRow(row: Model.SyslogEvent) {
+        try {
+            const text = JSON.stringify(row, null, 2);
+            await navigator.clipboard.writeText(text);
```
Thank you @MartinaeyNL, I was not aware of that. I will look into this.

sakib00 commented on ui/component/or-log-viewer/src/index.ts:
```diff
@@ -442,6 +452,18 @@ export class OrLogViewer extends translate(i18next)(LitElement) {
+        try {
+            const text = JSON.stringify(row, null, 2);
+            await navigator.clipboard.writeText(text);
+            // Optional: show some UI feedback
```
Thank you. I wasn't sure what to use, so I left a comment. I will update it accordingly.

sakib00 commented on ui/component/or-log-viewer/src/index.ts:
```diff
@@ -433,7 +434,16 @@ export class OrLogViewer extends translate(i18next)(LitElement) {
-                                    <td class="mdc-data-table__cell">${ev.message}</td>                                    
+                                    <td class="mdc-data-table__cell">${ev.message}</td>     
+                                    <td class="mdc-data-table__cell">
+                                        <button style="all:unset; cursor:pointer; display: flex; height: 1.5rem; width: 1.5rem;"
+                                              @click=${() => this._copyRow(ev)}
+                                              title="Copy this log to clipboard"
+                                        >
+                                        <svg xmlns="http://www.w3.org/2000/svg" fill="none" viewBox="0 0 24 24" stroke-width="1" stroke="currentColor" class="size-2">
+                                        <path stroke-linecap="round" stroke-linejoin="round" d="M15.75 17.25v3.375c0 .621-.504 1.125-1.125 1.125h-9.75a1.125 1.125 0 0 1-1.125-1.125V7.875c0-.621.504-1.125 1.125-1.125H6.75a9.06 9.06 0 0 1 1.5.124m7.5 10.376h3.375c.621 0 1.125-.504 1.125-1.125V11.25c0-4.46-3.243-8.161-7.5-8.876a9.06 9.06 0 0 0-1.5-.124H9.375c-.621 0-1.125.504-1.125 1.125v3.5m7.5 10. ...
+                                    </svg></button>
```
Thank you, I will update the code.

sakib00 commented on ui/component/or-log-viewer/src/index.ts:
```diff
@@ -422,7 +422,8 @@ export class OrLogViewer extends translate(i18next)(LitElement) {
                             <th style="width: 180px" class="mdc-data-table__header-cell" role="columnheader" scope="col">${i18next.t("subCategory")}</th>
-                            <th style="width: 100%; min-width: 300px;" class="mdc-data-table__header-cell" role="columnheader" scope="col">${i18next.t("message")}</th>                            
+                            <th style="width: 100%; min-width: 300px;" class="mdc-data-table__header-cell" role="columnheader" scope="col">${i18next.t("message")}</th>
+                            <th style="width: 100%; min-width: 300px;" class="mdc-data-table__header-cell" role="columnheader" scope="col"></th>
```
Thank you, yeah, it is my bad, I missed this. I will fix it.

sakib00 commented on ui/component/or-log-viewer/src/index.ts:
```diff
@@ -442,6 +452,18 @@ export class OrLogViewer extends translate(i18next)(LitElement) {
+    /** Copy a single log event to the clipboard as JSON */
+    protected async _copyRow(row: Model.SyslogEvent) {
+        try {
+            const text = JSON.stringify(row, null, 2);
```
Didn't the enhancement ask for an easier way for the user to copy the logs and paste it on the forum?
Or am I missing something? 😅

MartinaeyNL commented:
I'd like to check with @DonWillems what he thinks of these changes.

As a note from my end on the "Can't copy to clipboard";
When there is an error (like a insecure connection, or any other type), I think it's better to hide the copy button, right?
It'd be annoying that it would show "fail to copy" every time, while it isn't possible anyway.

> (even though I am not sure how I should test it...) 🤔
No problem. This code should practically catch all errors, so you can cause anything.
I don't remember if it failed for me using `localhost`, or only with a local IP address.
You can test with `npm run serve` to test `localhost`.

*Testing with local IP address is more difficult, as you need a custom setup using `OR_HOSTNAME` environment variable.*
*Both the Manager and Keycloak would need this environment variable, and you'd use [this](https://docs.openremote.io/docs/developer-guide/working-on-ui-and-apps#webpack-dev-server-environment-variables) to run the web dev server.*

sakib00 commented:
Hello @MartinaeyNL, I have updated the code, now the copy button shows only for secure connection. 

_How it looks on a **secure** connection:_
<img width="1413" alt="image" src="https://github.com/user-attachments/assets/5918cc22-e080-4e09-9b37-085440680c7b" />

_How it looks on a **not secure** connection:_
<img width="1461" alt="image" src="https://github.com/user-attachments/assets/7d0120cd-76a8-4f2e-8200-c7c8b027c9c6" />

DonWillems commented:
Hi @sakib00 ,

Thanks for your work. I'm sorry this issue wasn't more well-defined when you started. I have some comments:

1. The column doesn't have to be wider than 48px (so just give it that width). You can also consider adding the button to the end of the Message column cells?
2. Make sure the icon is still center aligned
3. Make the button `var(--or-app-color5, #CCCCCC));` if not hovered. Give the color is has now (`currentColor`/`--or-app-color4`) when the row is hovered (anywhere on the row). This will make the buttons less prominent on the page. To do this you might have to move away from or-mwc-input and use a plain button with or-icon to do this. You can look at the alarm button on the right in the Manager header for inspiration. 
4. Remove the `"` at the start and end of the message that you copy.

sakib00 commented:
Hello @DonWillems, thank you for your kind feedback. I have updated the points you have mentioned and it looks like this now. I have added `cursor:pointer` to the button but for some reason it is not showing in the video. 

https://github.com/user-attachments/assets/38398afa-571a-4d02-b896-e71e1ccb8e87

MartinaeyNL commented:
Hi @sakib00! I've updated your code with a few minor adjustments, so we can merge this now 💪 
Saves us time going back and forth haha

Merged! 🥳
